crates/wysiwyg/src/composer_model/replace_html.rs

```diff
@@ -0,0 +1,293 @@
+// Please see LICENSE in the repository root for full details.
+
+use regex::Regex;
+
+use crate::dom::html_source::HtmlSource;
+use crate::dom::nodes::ContainerNode;
+use crate::dom::parser::parse_from_source;
+use crate::{ComposerModel, ComposerUpdate, DomNode, Location, UnicodeString}; // Import the trait for to_tree
+
+impl<S> ComposerModel<S>
+where
+    S: UnicodeString,
+{
+    /// Replaces text in the current selection with new_html.
+    /// Treats its input as html that is parsed into a DomNode and inserted into
+    /// the document at the cursor.
+    pub fn replace_html(
+        &mut self,
+        new_html: S,
+        external_source: HtmlSource,
+    ) -> ComposerUpdate<S> {
+        self.push_state_to_history();
+        if self.has_selection() {
+            self.do_replace_text(S::default());
+        }
+        // Remove meta tags from the HTML which caused errors in html5ever
+        let meta_regex = Regex::new(r"<meta[^>]*>").unwrap();
+        let mut cleaned_html = meta_regex
+            .replace_all(&new_html.to_string(), "")
+            .to_string();
+
+        if external_source == HtmlSource::GoogleDoc {
+            // Strip outer b tag that google docs adds
+            let b_regex = Regex::new(r"<b[^>]*>(.*)<\/b>").unwrap();
+            cleaned_html = b_regex.replace(&cleaned_html, "$1").to_string();
+        }
+
+        let result =
+            parse_from_source(&cleaned_html.to_string(), external_source);
+
+        let doc_node = result.unwrap().into_document_node();
+        let (start, end) = self.safe_selection();
+        let range = self.state.dom.find_range(start, end);
+
+        // We should only have 1 dom node, so add the children under a paragraph to take advantage of the exisitng
+        // insert_node_at_cursor api and then delete the paragraph node promoting it's the children up a level.
+        let new_children = doc_node.into_container().unwrap().take_children();
+        let child_count = new_children.len();
+        let p = DomNode::Container(ContainerNode::new_paragraph(new_children));
+
+        let handle = self.state.dom.insert_node_at_cursor(&range, p);
+        self.state.dom.replace_node_with_its_children(&handle);
+        self.state.dom.wrap_inline_nodes_into_paragraphs_if_needed(
+            &self.state.dom.parent(&handle).handle(),
+        );
+
+        // Track the index of the last inserted node for placing the cursor
+        let last_index = handle.index_in_parent() + child_count - 1;
+        let last_handle = handle.parent_handle().child_handle(last_index);
+        let location = self.state.dom.location_for_node(&last_handle);
+
+        self.state.start =
+            Location::from(location.position + location.length - 1);
+        self.state.end = self.state.start;
+        // add a trailing space in cases when we do not have a next sibling
+        self.create_update_replace_all()
+    }
+}
+
+#[cfg(test)]
```

Comment:
Can we have some comments stating which structure these huge HTML chunks should represent? It's quite difficult to understand if the tests are doing the right thing since we can't compare the tree structures.

Reply:
Yea most of this chunk is actually the same as the one in `parse.rs` so I'll use format here to make things clearer and avoid the copy paste and add some comments.